src/dune_cache/local.ml

```diff
@@ -43,37 +43,6 @@ module Target = struct
   ;;
 end
 
-(* This function is like [Unix.link] but handles the "Too many links" error by
-   creating a copy of the [src] in a temporary directory, then atomically
-   replacing the [src] with the copy, and finally creating the requested [dst]
-   by calling [Unix.link src dst] again.
-
-   We hit the "Too many links" error because we store a lot of empty files in
-   the cache, which all get deduplicated into the same cache entry. This
-   function essentially deletes the "overlinked" entry from the cache, creating
-   a fresh copy with the 0 link count. This leads to some duplication but it's
-   negligible: we might store the empty file several times across all workspaces
-   instead of just storing it once.
-
-   If you need to debug this function, you can trigger the "Too many links"
-   error by running [for i in {1..100000}; do ln $file tmp/$i; done], where the
-   [$file] is the shared cache entry for the empty file. After that, no more
-   hard links on [$file] will be allowed, triggering the [EMLINK] code path. *)
-let link_even_if_there_are_too_many_links_already ~src ~dst =
-  try Path.link src dst with
-  | Unix.Unix_error (Unix.EMLINK, _, _)
-  | Unix.Unix_error (Unix.EUNKNOWNERR -1142, _, _) (* Needed for OCaml < 5.1 *) ->
-    Temp.with_temp_file ~dir:temp_dir ~prefix:"dune" ~suffix:"copy" ~f:(function
-      | Error e -> raise e
-      | Ok temp_file ->
-        Io.copy_file ~src ~dst:temp_file ();
-        (* This replaces [src], which has too many links already, with a fresh
-           copy we've just created in the [temp_file]. *)
-        Path.rename temp_file src;
-        (* This should now succeed. *)
-        Path.link src dst)
```

Comment:
The difference between `Io.portable_hardlink` is that it will just copy when hard linking fails rather than copying and then linking.

Reply:
If you read the above comment carefully, you'll see that the old code was doing more than just falling back to copying. I don't think this place needs to change because it's not really broken. You are replacing it with something simpler but also less optimal.

Reply:
Sure, this function allows for more sharing. But the only files where we get "too many links" are small anyway. So I didn't think it was worth it. I could restore it though.

Reply:
We can discuss simplifying this code (you may be right that it's not worth it) but it seems wrong to conflate a bug fix and a possibly controversial change that removes an optimisation in the same PR.